src/client/AppEmbeddedSDK.ts

```diff
@@ -0,0 +1,63 @@
+import { randomString } from './utils';
+
+const ACTION_ID_LENFTH = 80;
+
+export class AppEmbeddedSDK {
+    private listener: (this: Window, ev: MessageEvent) => any;
+    private callbacks: Map<string, (response: any) => any>;
+
+    constructor() {
+        this.listener = () => console.log('init');
+        this.callbacks = new Map();
+    }
+
+    public getUserInfo(): Promise<any> {
+        return this.call(AppEmbeddedSDKActions.GET_USER_INFO);
+    }
+
+    public getRoomInfo(): Promise<any> {
+        return this.call(AppEmbeddedSDKActions.GET_ROOM_INFO);
+    }
+
+    /**
+     * Create a message listener with id and add them to callbacks map set.
+     * @param id the id of the message event listener
+     */
+    private initListener(): void {
+        this.listener = ({ data }) => {
+            if (!data.hasOwnProperty('rcEmbeddedSDK')) {
+                return;
+            }
+
+            const { rcEmbeddedSDK: { id, payload } } = data;
+
+            if (this.callbacks.has(id)) {
+                const resolve = this.callbacks.get(id);
+
+                if (typeof resolve === 'function') {
+                    resolve(payload);
+                }
+                this.callbacks.delete(id);
+            }
+        };
+        window.addEventListener('message', this.listener);
+    }
+    private call(action: string, payload?: any ): Promise<any>  {
+        return new Promise((resolve) => {
+            const id = randomString(ACTION_ID_LENFTH);
+
+            this.initListener();
+
+            if (window.parent) {
```

Comment:
Funny thing is `window.parent` is always available - if the window has no actual parent, the `window.parent` property is only a reference to itself :) So you don't need to check this.

[As stated on MDN](https://developer.mozilla.org/en-US/docs/Web/API/Window/parent)

Reply:
Thanks!